### PR TITLE
Add cleanup function to remove all the objects from a bucket after tests

### PR DESCRIPTION
**Pull request by nitisht**
Fixes : https://github.com/minio/mint/issues/165
and improves logging by adding args in few test cases.

### CODE REVIEW

File: functional_tests.go [hunk 1]
```diff
@@ -379,9 +411,9 @@ func testMakeBucketRegions() {
 		failureLog(function, args, startTime, "", "MakeBucket failed", err).Fatal()
 	}
 
-	// Remove the newly created bucket.
-	if err = c.RemoveBucket(bucketName + ".withperiod"); err != nil {
-		failureLog(function, args, startTime, "", "Remove bucket failed", err).Fatal()
+	// cleanup
```
Comment (harshavardhana) [inline functional_tests.go]: Add more descriptive comment rather than place holder like `cleanup`

File: functional_tests.go [hunk 2]
```diff
@@ -1613,7 +1608,7 @@ func testPutObjectWithContext() {
 	args := map[string]interface{}{
 		"bucketName": "",
 		"objectName": "",
-		"opts":       "minio.PutObjectOptions{ContentType:objectContentType}",
```
Comment (harshavardhana) [inline functional_tests.go]: why this change? shouldn't we add this?

Comment (nitisht) [reply functional_tests.go]: was a typo, fixed.

File: functional_tests.go (hunk 1 above)
Comment (nitisht) [reply functional_tests.go]: Agreed. Updated the comments

File: functional_tests.go [hunk 3]
```diff
@@ -121,6 +121,31 @@ func ignoredLog(function string, args map[string]interface{}, startTime time.Tim
...
+// Delete objects and buckets recursively
+func cleanupBuckets(c *minio.Client) (err error) {
+
+	doneCh := make(chan struct{})
+	defer close(doneCh)
+
+	buckets, err := c.ListBuckets()
+
+	for _, bucket := range buckets {
+		for obj := range c.ListObjects(bucket.Name, "", true, doneCh) {
```
Comment (harshavardhana) [inline functional_tests.go]: you can put each ListObjects() on a go-routine since they are independent buckets and clean them up.

File: functional_tests.go [hunk 4]
```diff
@@ -121,6 +121,37 @@ func ignoredLog(function string, args map[string]interface{}, startTime time.Tim
...
+					}
+				} else {
+					err = object.Err
+				}
+			}
+		}(bucket.Name)
+	}
+	return err
+}
+
```
[the remark this answers is not in the document]
Comment (nitisht) [reply functional_tests.go]: Yes it will delete everything on the server. Let me add the bucket prefix check

File: functional_tests.go [hunk 5]
```diff
@@ -121,6 +121,37 @@ func ignoredLog(function string, args map[string]interface{}, startTime time.Tim
...
+	var err error
+	var buckets []minio.BucketInfo
+
+	// Create a done channel to control 'ListObjectsV2' go routine.
+	doneCh := make(chan struct{})
+
+	// Indicate to our routine to exit cleanly upon return.
+	defer close(doneCh)
+
+	buckets, err = c.ListBuckets()
```
Comment (harshavardhana) [inline functional_tests.go]: Return on error right here..

File: functional_tests.go [hunk 6]
```diff
@@ -121,6 +121,37 @@ func ignoredLog(function string, args map[string]interface{}, startTime time.Tim
...
+	for _, bucket := range buckets {
+		go func(bucketName string) {
+			objectCh := c.ListObjectsV2(bucketName, "", true, doneCh)
+			for object := range objectCh {
+				if object.Err != nil {
+					if object.Key != "" {
+						err = c.RemoveObject(bucketName, object.Key)
+					}
+				} else {
+					err = object.Err
```
Comment (harshavardhana) [inline functional_tests.go]: Do not set the err from outside scope in a go-routine,  collect all errors.. in a errs []error and then store them based on the buckets index. like this 

```go
for i, bucket := range buckets {
     go func(i int, bucketName string) {
           ....
           errs[i] = err
     }(i, bucket)
}
```

Comment (harshavardhana) [reply functional_tests.go]: Also since there is no wait group, this won't delete all buckets all the time. Since the program might exit if this function returns after all go-routines have spawned. 

Use wait group to limit yourself and use `wg.Wait()` to wait for cleaning up of all buckets. Once done loop through 

```go
for _, err := range errs {
        if err != nil {
             return err 
        }
}
```

**Comment by nitisht** [comment]
Updated the PR with a simpler approach to cleanup, since previous approach failed if there are multiple instances of `minio-go` tests running in parallel against a single server, `cleanup` deleted buckets belonging to other test instances as well - for example when travis runs multiple `minio-go` tests against play server. 

Please take a look @harshavardhana @poornas

File: functional_tests.go [hunk 7]
```diff
@@ -121,6 +121,33 @@ func ignoredLog(function string, args map[string]interface{}, startTime time.Tim
...
+			}
+		}
+	}
+	// objects are already deleted, clear the buckets now
+	err := c.RemoveBucket(bucketName)
+	if err != nil {
+		return err
+	}
+	return err
+}
```
Comment (poornas) [inline functional_tests.go]: what if there are incomplete upload parts? that would need cleanup as well.

Comment (nitisht) [reply functional_tests.go]: can you explain a bit more @poornas

Comment (poornas) [reply functional_tests.go]: if there are incomplete uploads to a bucket, RemoveBucket would fail. For tests using large uploads, shouldn't the cleanup also remove incomplete uploads if any?

Comment (nitisht) [reply functional_tests.go]: done @poornas

Comment (harshavardhana) [reply functional_tests.go]: RemoveBucket works fine on all gateways even with incomplete uploads, that is the S3 behavior. But it is a good change since it will cater for B2 gateway backend as well. Which requires explicit removal of incomplete uploads.

**Comment by harshavardhana** [comment]
> Partially fixes : minio/mint#165
> and improves logging by adding args in few test cases.

Is this true that it partially fixes? @nitisht

**Comment by nitisht** [comment]
> Is this true that it partially fixes?

Let me change it to fixes, initially there were stale buckets left by minio-js as well

```
[2017-09-29 17:11:02 PDT] 100KiB 41b17354-ff91-478d-9a46-353830d48767/datafile-100-kB
[2017-09-29 17:11:02 PDT] 100KiB 41b17354-ff91-478d-9a46-353830d48767/datafile-100-kB.buffer
```

but that is not seen anymore.
